ec/src/models/bw6/mod.rs

```diff
@@ -185,113 +211,101 @@ impl<P: BW6Config> BW6<P> {
+            let c = Self::exp_by_x_minus_1_div_3(&b);
+            // D = C**(u-1)
+            let d = Self::exp_by_x_minus_1(&c);
+            // E = (D**(u-1))**(u-1) * D
+            let e = Self::exp_by_x_minus_1(&Self::exp_by_x_minus_1(&d)) * &d;
+            // F = (E**(u+1) * C).conjugate() * D
+            let f = (Self::exp_by_x_plus_1(&e) * &c)
+                .cyclotomic_inverse()
+                .unwrap()
+                * &d;
+            // G = ((F * D)**(u+1)).conjugate() * C * B
+            let g = Self::exp_by_x_plus_1(&(f * &d))
+                .cyclotomic_inverse()
+                .unwrap()
+                * &c
+                * &b;
+            // d2 = (ht**2+3*hy**2)//4
+            let d2 = ((P::H_T * P::H_T + 3 * P::H_Y * P::H_Y) / 4) as u64;
+            // d1 = (ht-hy)//2
+            let d1 = (P::H_T - P::H_Y) / 2;
+            // H = F**d1 * E
+            let mut h = f.cyclotomic_exp(&[d1 as u64]);
+            if d1 < 0 {
+                h.cyclotomic_inverse_in_place();
+            }
+            h *= e;
+            // H = H**2 * H * B * G**d2
+            let h = h.square() * &h * &b * g.cyclotomic_exp(&[d2]);
+            // return A * H
+            a * &h
+        } else {
+            // Algorithm 4.4 from https://yelhousni.github.io/phd.pdf
+            // Follows the implementation https://gitlab.inria.fr/zk-curves/snark-2-chains/-/blob/master/sage/pairing_bw6_bls12.py#L969
+
+            // A = m**(u-1)
+            let a = Self::exp_by_x_minus_1(f);
+            // A = A**(u-1)
+            let a = Self::exp_by_x_minus_1(&a);
+            // A = A * m.frobenius()
+            let a = a * f.frobenius_map(1);
+            // B = A**(u+1) * m.conjugate()
+            let b = Self::exp_by_x_plus_1(&a) * f.cyclotomic_inverse().unwrap();
+            // A = A**2 * A
+            let a = a.square() * &a;
+            // C = B**((u-1)//3)
+            let c = Self::exp_by_x_minus_1_div_3(&b);
+            // D = C**(u-1)
+            let d = Self::exp_by_x_minus_1(&c);
+            // E = (D**(u-1))**(u-1) * D
+            let e = Self::exp_by_x_minus_1(&Self::exp_by_x_minus_1(&d)) * &d;
+            // D = D.conjugate()
+            let d = d.cyclotomic_inverse().unwrap();
+            // Fc = D * B
+            let fc = d * &b;
+            // G = E**(u+1) * Fc
+            let g = Self::exp_by_x_plus_1(&e) * &fc;
+            // H = G * C
+            let h = g * &c;
+            // I = (G * D)**(u+1) * Fc.conjugate()
+            let i = Self::exp_by_x_plus_1(&(g * &d)) * fc.cyclotomic_inverse().unwrap();
+            // d2 = (ht**2+3*hy**2)//4
+            let d2 = ((P::H_T * P::H_T + 3 * P::H_Y * P::H_Y) / 4) as u64;
+            // d1 = (ht+hy)//2
+            let d1 = (P::H_T + P::H_Y) / 2;
+            // J = H**d1 * E
+            let mut j = h.cyclotomic_exp(&[d1 as u64]);
+            if d1 < 0 {
+                j.cyclotomic_inverse_in_place();
+            }
```

Comment:
If these are the only places where we use `P::H_T` and `P::H_Y`, it might be worth it to wrap the computations of `d2` and `d1` into helper methods where we can additionally enforce  

(a) that the cast to `u64` works correctly (i.e. via a `try_from`),
(b) that we're correctly handling the d1 < 0 case.

Reply:
So far `H_T` and `H_Y`  are used only in the final exp, but might appear useful later. Otherwise we could just hardcode these `d1` and `d2` to the Config.

I addressed handling the `d1<0` case, by adding `cyclotomic_exp_signed()`. As for `try_from`, I think `as u64` does exactly what we want for negative `d1`s.